File: hydra-python/src/main/python/hydra/lib/chars.py
"""Python implementations of hydra.lib.chars primitives."""


def is_alpha_num(value: int) -> bool:
    """Check if a character (as int) is alphanumeric."""
    return chr(value).isalnum()


def is_lower(value: int) -> bool:
    """Check if a character (as int) is lowercase."""
    return chr(value).islower()


def is_space(value: int) -> bool:
    """Check if a character (as int) is whitespace."""
    return chr(value).isspace()


def is_upper(value: int) -> bool:
    """Check if a character (as int) is uppercase."""
    return chr(value).isupper()


def to_lower(value: int) -> int:
    """Convert a character (as int) to lowercase, returning the int value."""
    return ord(chr(value).lower())


def to_upper(value: int) -> int:
    """Convert a character (as int) to uppercase, returning the int value."""
    return ord(chr(value).upper())

File: hydra-haskell/src/main/haskell/Hydra/Lib/Chars.hs
-- | Haskell implementations of hydra.lib.chars primitives.
-- All functions operate on Unicode code points (Int) rather than Char.

module Hydra.Lib.Chars where

import qualified Data.Char as C


-- | Check whether a character is alphanumeric.
isAlphaNum :: Int -> Bool
isAlphaNum = C.isAlphaNum . C.chr

-- | Check whether a character is lowercase.
isLower :: Int -> Bool
isLower = C.isLower . C.chr

-- | Check whether a character is a whitespace character.
isSpace :: Int -> Bool
isSpace = C.isSpace . C.chr

-- | Check whether a character is uppercase.
isUpper :: Int -> Bool
isUpper = C.isUpper . C.chr

-- | Convert a character to lowercase.
toLower :: Int -> Int
toLower = C.ord . C.toLower . C.chr

-- | Convert a character to uppercase.
toUpper :: Int -> Int
toUpper = C.ord . C.toUpper . C.chr
